src/Datastore/DatastoreClient.php

```diff
@@ -0,0 +1,907 @@
+    }
+
+    /**
+     * Create an entity.
+     *
+     * This method does not execute any service requests.
+     *
+     * ```
+     * $key = $datastore->key('Person', 'Bob');
+     * $entity = $datastore->entity($key, [
+     *     'firstName' => 'Bob',
+     *     'lastName' => 'Testguy'
+     * ]);
+     * ```
+     *
+     * ```
+     * // Entities can be manipulated via PHP's arrayaccess:
+     * $entity['firstName'] = 'Bob';
+     * $entity['lastName'] = 'Testguy';
+     *
+     * echo $entity['firstName']; // 'Bob'
+     * ```
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1beta3/Entity Entity
+     *
+     * @param Key $key The Key used to uniquely identify this entity. Only
+     *        complete keys may be used.
+     * @param array $entity The data to fill the entity with.
+     * @param array $options See {@see Google\Cloud\Datastore\Entity::__construct()} for documentation.
+     * @return Entity
+     */
+    public function entity(Key $key, array $entity = [], array $options = [])
+    {
+        return new Entity($key, $entity, $options);
+    }
+
+    /**
+     * Allocates an available ID to a given incomplete key
+     *
+     * Key MUST be in an incomplete state (i.e. including a kind but not an ID
+     * or name in its final pathElement).
+     *
+     * This method will execute a service request.
+     *
+     * Example:
+     * ```
+     * $key = $datastore->key('Person');
+     * $allocatedId = $datastore->allocateId($key);
+     * ```
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1beta3/projects/allocateIds allocateIds
+     *
+     * @param Key $key The incomplete key
+     * @param array $options Configuration options
+     * @return Key
+     */
+    public function allocateId(Key $key, array $options = [])
+    {
+        $res = $this->allocateIds([$key], $options);
+        return $res[0];
+    }
+
+    /**
+     * Allocate available IDs to a set of keys
+     *
+     * Keys MUST be in an incomplete state (i.e. including a kind but not an ID
+     * or name in their final pathElement).
+     *
+     * This method will execute a service request.
+     *
+     * Example:
+     * ```
+     * $keys = $datastore->keys('Person', ['number' => 2]);
+     * $allocatedIds = $datastore->allocateIds($keys);
+     * ```
+     *
+     * @see https://cloud.google.com/datastore/reference/rest/v1beta3/projects/allocateIds allocateIds
+     *
+     * @param Key[] $keys The incomplete keys
+     * @param array $options Configuration options
+     * @return Key[]
+     */
+    public function allocateIds(array $keys, array $options = [])
+    {
+        // Validate the given keys. First check types, then state of each.
+        // The API will throw a 400 if the key is complete, but it's an easy
+        // check we can handle before going to the API to save a request.
+        // @todo replace with json schema
+        $this->validateBatch($keys, Key::class, function ($key) {
```

Comment:
Would it ever be a valid use case to have an array containing complete and incomplete keys? Would it be possible to just allocateIds to the keys in that array that are incomplete?